src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,1154 @@
+
+    function collectReadsAndWrites(
+        targetRange: TargetRange,
+        scopes: Scope[],
+        enclosingTextRange: TextRange,
+        sourceFile: SourceFile,
+        checker: TypeChecker) {
+
+        const usagesPerScope: ScopeUsages[] = [];
+        const substitutionsPerScope: Map<Node>[] = [];
+        const errorsPerScope: Diagnostic[][] = [];
+        const visibleDeclarationsInExtractedRange: Symbol[] = [];
+
+        // initialize results
+        for (const _ of scopes) {
+            usagesPerScope.push({ usages: createMap<UsageEntry>(), substitutions: createMap<Expression>() });
+            substitutionsPerScope.push(createMap<Expression>());
+            errorsPerScope.push([]);
+        }
+        const seenUsages = createMap<Usage>();
+        let valueUsage = Usage.Read;
+        const target = isReadonlyArray(targetRange.range) ? createBlock(<Statement[]>targetRange.range) : targetRange.range;
+        const containingLexicalScopeOfExtraction = isBlockScope(scopes[0], scopes[0].parent) ? scopes[0] : getEnclosingBlockScopeContainer(scopes[0]);
+
+        collectUsages(target);
+
+        for (let i = 0; i < scopes.length; i++) {
+            let hasWrite = false;
+            let readonlyClassPropertyWrite: Declaration | undefined = undefined;
+            usagesPerScope[i].usages.forEach(value => {
+                if (value.usage === Usage.Write) {
+                    hasWrite = true;
+                    if (value.symbol.flags & SymbolFlags.ClassMember &&
+                        value.symbol.valueDeclaration &&
+                        hasModifier(value.symbol.valueDeclaration, ModifierFlags.Readonly)) {
+                        readonlyClassPropertyWrite = value.symbol.valueDeclaration;
+                    }
+                }
+            });
+
+            if (hasWrite && !isArray(targetRange.range) && isExpression(targetRange.range)) {
+                errorsPerScope[i].push(createDiagnosticForNode(targetRange.range, Messages.CannotCombineWritesAndReturns));
+            }
+            else if (readonlyClassPropertyWrite && i > 0) {
+                errorsPerScope[i].push(createDiagnosticForNode(readonlyClassPropertyWrite, Messages.CannotCombineWritesAndReturns));
+            }
+        }
+
+        // If there are any declarations in the extracted block that are used in the same enclosing
+        // lexical scope, we can't move the extraction "up" as those declarations will become unreachable
+        if (visibleDeclarationsInExtractedRange.length) {
+            forEachChild(containingLexicalScopeOfExtraction, checkForUsedDeclarations);
+        }
+
+        return { target, usagesPerScope, errorsPerScope };
+
+        function collectUsages(node: Node) {
+            if (isDeclaration(node) && node.symbol) {
+                visibleDeclarationsInExtractedRange.push(node.symbol);
+            }
+
+            if (isAssignmentExpression(node)) {
+                const savedValueUsage = valueUsage;

```

Comment:
>valueUsage [](start = 40, length = 10)

Personally, I would just make `valueUsage` a parameter of `collectUsages`.  That would eliminate both the side effects and the saving/restoration.